File: aoc2018/aoc2018-java/src/main/java/de/havox_design/aoc2018/day22/ModeMaze.java
package de.havox_design.aoc2018.day22;

import de.havox_design.aoc.utils.java.AoCFunctionality;
import de.havox_design.aoc.utils.kotlin.model.positions.Position2d;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class ModeMaze implements AoCFunctionality {
    private static final int ID_INPUT_DEPTH = 0;
    private static final int ID_INPUT_TARGET = 1;

    private final int depth;
    private final Position2d<Integer> target;
    private final Map<Position2d<Integer>, Double> erosionLevels;

    public ModeMaze(String fileName) {
        List<String> input = readData(fileName);
        final DepthParser depthParser = new DepthParser();
        final TargetParser targetParser = new TargetParser();

        this.depth = depthParser.parse(input.get(ID_INPUT_DEPTH));
        this.target = targetParser.parse(input.get(ID_INPUT_TARGET));
        this.erosionLevels = new HashMap<>();
    }

    public static long processTask1(String fileName) {
        ModeMaze instance = new ModeMaze(fileName);
        return instance.processTask1();
    }

    public static long processTask2(String fileName) {
        ModeMaze instance = new ModeMaze(fileName);
        return instance.processTask2();
    }

    public long processTask1() {
        int[][] cave = new int[target.getY() + 1][target.getX() + 1];

        int riskLevel = 0;

        for (int y = 0; y < cave.length; y++) {
            String row = y + ":\t";
            for (int x = 0; x < cave[0].length; x++) {
                int type = getType(x, y);

                if (type == 0) {
                    row += ".";
                } else if (type == 1) {
                    row += "=";
                } else if (type == 2) {
                    row += "|";
                }

                riskLevel += type;
            }
            //System.out.println(row);
        }

        return riskLevel;
    }

    public long processTask2() {
        return 0;
    }

    private int getType(int x, int y) {
        Position2d<Integer> key = new Position2d<>(x, y);
        double value;
        if ((x == 0 && y == 0) || (y == target.getY() && x == target.getX())) {
            value = 0D;
        } else if (y == 0) {
            value = x * 16807D;
        } else if (x == 0) {
            value = y * 48271D;
        } else {
            double left = erosionLevels.get(new Position2d<>(x - 1, y));
            double top = erosionLevels.get(new Position2d<>(x, y - 1));
            value = left * top;
        }
        double erosionLevel = (value + depth) % 20183;
        erosionLevels.put(key, erosionLevel);

        return ((int) erosionLevel) % 3;
    }

}

File: aoc2018/aoc2018-kotlin/src/main/kotlin/de/havox_design/aoc2018/day22/ModeMaze.kt
package de.havox_design.aoc2018.day22

import de.havox_design.aoc.utils.kotlin.model.positions.Position2d
import java.util.PriorityQueue

class ModeMaze(private var filename: String) {
    fun processTask1(): Any {
        val data = getData()
        val depth = data.first
        val target = data.second

        return getRiskLevel(depth, target)
    }

    fun processTask2(): Any {
        val data = getData()
        val depth = data.first
        val target = data.second

        return getShortestTimeTo(depth, target)
    }

    private fun getRiskLevel(depth: Int, target: Position2d<Int>) =
        HardCave(depth, target)
            .apply {
                (0..target.y)
                    .forEach { y ->
                        (0..target.x)
                            .forEach { x ->
                                at(Position2d(x, y))
                            }
                    }
            }
            .getTotalRiskRiskLevel()

    private fun getShortestTimeTo(depth: Int, target: Position2d<Int>): Int {
        val cave = HardCave(depth, target)
        val start = cave
            .at(Position2d(0, 0))
            .let { CaveStep(it, it.toolsRequired().first()) }
        val seenShortest = mutableMapOf(start.asKey to start.cost)
        val stepsRemaining = PriorityQueue<CaveStep>().apply { add(start) }

        while (stepsRemaining.isNotEmpty()) {
            val step = stepsRemaining.poll()

            if (step.at.point == target && step.using == Tool.TORCH) {
                return step.cost
            }

            step
                .generateSteps(cave)
                .forEach {
                    val existing = seenShortest[it.asKey]

                    if (existing == null || it.cost < existing) {
                        seenShortest[it.asKey] = it.cost
                        stepsRemaining += it
                    }
                }
        }

        return -1
    }

    private fun getData(): Pair<Int, Position2d<Int>> {
        val input = getResourceAsText(filename)

        val depth = input[0].split(':')[1].trim().toInt()
        val target = Position2d(
            input[1].split(':')[1].trim().split(',')[0].trim().toInt(),
            input[1].split(':')[1].trim().split(',')[1].trim().toInt()
        )

        return Pair(depth, target)
    }

    private fun getResourceAsText(path: String): List<String> =
        this
            .javaClass
            .classLoader
            .getResourceAsStream(path)!!
            .bufferedReader()
            .readLines()
}
